query.go

```diff
@@ -778,120 +772,272 @@ func (s *ChainService) GetCFilter(blockHash chainhash.Hash,
-				// know about it and we declare it sane. We can
-				// kill the query and pass the response back to
-				// the caller.
-				filter = gotFilter
-				close(quit)
-			default:
-			}
-		},
-		options...,
-	)
+					if gotHeader != curHeader {
+						return
+					}
+
+					// At this point, the filter matches
+					// what we know about it and we declare
+					// it sane. If this is the filter
+					// requested initially, send it to the
+					// caller immediately.
+					if i == 1 {
+						filterChan <- gotFilter
+					}
+
+					// Put the filter in the cache and
+					// persistToDisk if the caller
+					// requested it.
+					// TODO(halseth): for an LRU we could
+					// take care to insert the next height
+					// filter last.
+					evict, err := s.putFilterToCache(
+						&response.BlockHash,
+						dbFilterType, gotFilter,
+					)
+					if err != nil {
+						log.Warnf("Couldn't write "+
+							"filter to cache: %v",
+							err)
+					}
+
+
+					// TODO(halseth): dynamically
+					// increase/decrease the batch size to
+					// match our cache capacity.
+					if evict && s.FilterCache.Len() < int(numFilters) {
+						log.Warnf("Items evicted from "+
+							"the cache with less "+
+							"than %d elements. "+
+							"Consider increasing "+
+							"the cache size...",
+							numFilters)
+					}
+
+					if qo.persistToDisk {
+						err = s.FilterDB.PutFilter(
+							&blockHash, filter,
+							dbFilterType,
+						)
+						if err != nil {
+							log.Warnf("Couldn't "+
+								"write filter "+
+								"to filterDB: "+
+								"%v", err)
+						}
+
+						log.Tracef("Wrote filter for "+
+							"block %s, type %d",
+							blockHash, filterType)
+					}
 
-	if filter != nil {
-		// If we found a filter, put it in the cache and persistToDisk if
-		// the caller requested it.
-		_, err := s.putFilterToCache(&blockHash, dbFilterType, filter)
-		if err != nil {
-			log.Warnf("couldn't write filter to cache: %v", err)
+					// Finally, we can delete it from the
+					// heightIndex.
+					delete(heightIndex, response.BlockHash)
+
+					// If the heightIndex is empty, we got
+					// everything we wanted, and can exit.
+					if len(heightIndex) == 0 {
+						close(quit)
+					}
+				default:
+				}
+			},
+			options...,
+		)
+
+		// If there are elements left to receive, the query failed.
+		if len(heightIndex) > 0 {
+			log.Errorf("Query failed with %d out of %d filters "+
+				"received", len(heightIndex), numFilters)
+			return
 		}
+	}()
 
-		qo := defaultQueryOptions()
-		qo.applyQueryOptions(options...)
-		if qo.persistToDisk {
-			err := s.FilterDB.PutFilter(&blockHash, filter, dbFilterType)
-			if err != nil {
-				return nil, err
-			}
+	var ok bool
+	select {
 
-			log.Tracef("Wrote filter for block %s, type %d",
-				blockHash, filterType)
+	// We'll return immediately to the caller when the filter arrives.
+	case filter, ok = <-filterChan:
+		if !ok {
+			// TODO: return error?
+			return nil, nil
```

Comment:
We can use an error chan to send any of the fatal errors back and use that here.

Reply:
Just returning nil for now, to not change the behavior of `GetCFilter` from the perspective of the caller.